### PR TITLE
Add tests for `mixed_precision` + `variables`

### PR DESCRIPTION
**Pull request by Faisal-Alsrheed**
Changes:

1. Added a new test file, `dtype_policy_test.py`.
2. Added new tests to `variables_test.py`

### CODE REVIEW

File: keras/backend/common/variables_test.py [hunk 1]
```diff
@@ -1,15 +1,22 @@
...
 from keras.backend.common.variables import KerasVariable
+from keras.backend.common.variables import shape_equal
+from keras.backend.common.variables import standardize_dtype
 from keras.backend.common.variables import standardize_shape
 from keras.testing import test_case
 
 
-class VariablesTest(test_case.TestCase):
+class VariableInitializationTest(test_case.TestCase):
+    """tests for lines unders KerasVariable __init__ method"""
```
Comment (fchollet) [inline keras/backend/common/variables_test.py:16]: `Tests for KerasVariable.__init__()`

File: keras/callbacks/model_checkpoint_test.py [hunk 2]
```diff
@@ -448,6 +448,12 @@ def get_model():
     )
     @pytest.mark.requires_trainable_backend
     def test_model_checkpoint_loading(self):
+        # TODO This test sometimes FAILED for Torch and Jax
```
Comment (fchollet) [inline keras/callbacks/model_checkpoint_test.py:451]: We've investigated this a bit but couldn't identify the cause.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras/callbacks/model_checkpoint_test.py (hunk 2 above)
Comment (Faisal-Alsrheed) [reply keras/callbacks/model_checkpoint_test.py:451]: I will do this task tomorrow.
I will make small unit tests until I find the bug. I enjoy that :)

File: keras/backend/common/variables_test.py [hunk 3]
```diff
@@ -128,15 +185,716 @@ def test_standardize_shape_with_valid_input(self):
...
+    # def test_standardize_shape_with_torch_Size_containing_float(self):
+    #     """Tests shape with a float value."""
+    #     shape_with_float = (3, 4, 5.0)
+    #     with self.assertRaisesRegex(
+    #         ValueError,
+    #         "Cannot convert .* to a shape. Found invalid entry '5.0'.",
+    #     ):
+    #         _ = standardize_shape(shape_with_float)
+
+    def test_standardize_shape_with_torch_Size_valid(self):
```
Comment (fchollet) [inline keras/backend/common/variables_test.py:767]: Please keep test method names as snake_case

File: keras/backend/common/variables_test.py [hunk 4]
```diff
@@ -128,15 +185,716 @@ def test_standardize_shape_with_valid_input(self):
...
+
+    def test_shape_equal_length_mismatch(self):
+        """Test mismatch in lengths of shapes."""
+        self.assertFalse(shape_equal((3, 2), (3, 2, 4)))
+        self.assertFalse(shape_equal((), (3,)))
+        self.assertFalse(shape_equal((3, 2, 4, 5), (3, 2, 4)))
+
+    # def test_variable_rand(self):
+    #     """Test reverse & operation on a variable."""
+    #     # TODO
```
Comment (fchollet) [inline keras/backend/common/variables_test.py:611]: Are these intended to be addressed in this PR, or later?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

**Comment by Faisal-Alsrheed** [comment]
Hi @fchollet, 

I believe now is the perfect time to gather your feedback on how you'd like me to organize my contributions. 

I'm committed to increasing the test coverage to 100% and am really enjoying it.

To be honest, only two days ago, I learned how to work on multiple PRs without any conflicts between them.

But I've always wondered: Do you prefer short PRs or long ones? I'm new to GitHub, but I believe each PR should focus on one specific topic, correct?

I suggest that, moving forward, I begin with testing fundamental functions, like those in the backend and utility folders, ensuring they are 100% covered, one PR for one test file a time . After that, I'll move on to higher-level functions. If I discover a bug, I'll attempt to fix it if it's straightforward. If it's more complex, would you like me to open an issue in such cases? 


Please feel free to be very tough with your code feedback; I'll learn more this way. I'm here to learn so I can teach my son in three years :)

Thank you.

File: keras/backend/common/variables_test.py (hunk 4 above)
Comment (Faisal-Alsrheed) [reply keras/backend/common/variables_test.py:632]: I think it is better to do it in this PR.

**Comment by Faisal-Alsrheed** [comment]
@fchollet 
Please see the failed tests and give me your feedback how to fix them.
Thank you.

**Comment by fchollet** [comment]
Thanks for the update!

> def test__pos__(self):

I think we can replace the implementation of `def __pos__` in `keras/backend/common/variables.py:Variable` with just `return self.value`. Is there any context in which this would be incorrect?

> def test__div__(self):

If this only occurs for the numpy backend, we can override `def __div__` in `keras/backend/numpy/core.py:Variable` to use an implementation relying on `np.divide`.

> __round__

If it's just TF, we can likewise override `__round__` on the TF Variable class and use `tf.math.round`.

> def test__lshift__(self):

Let's just outright remove all `*shift` methods on `keras/backend/common/variables.py:Variable` and remove the tests. They're not very useful.

**Comment by Faisal-Alsrheed** [comment]
@fchollet

I have removed all *shift methods and tests.

I have kept the others to see the test results.

Please check the failed test summary for each backend. We got different results. I need your feedback. 

Thank you.

**Comment by Faisal-Alsrheed** [comment]
> Thanks for the update!
> 
> > def test__pos__(self):
> 
> I think we can replace the implementation of `def __pos__` in `keras/backend/common/variables.py:Variable` with just `return self.value`. Is there any context in which this would be incorrect?
> 
> > def test__div__(self):
> 
> If this only occurs for the numpy backend, we can override `def __div__` in `keras/backend/numpy/core.py:Variable` to use an implementation relying on `np.divide`.
> 
> > **round**
> 
> If it's just TF, we can likewise override `__round__` on the TF Variable class and use `tf.math.round`.
> 
> > def test__lshift__(self):
> 
> Let's just outright remove all `*shift` methods on `keras/backend/common/variables.py:Variable` and remove the tests. They're not very useful.

**Comment by Faisal-Alsrheed** [comment]
@fchollet

I have removed all *shift methods and tests.

I've retained the others to see the test results.

Please check the failed test summary for each backend. We got different results, and I need your feedback.

Thank you.

*Sorry, I closed it by mistake; I have reopened it again.

**Comment by fchollet** [comment]
Thanks for the update! You can fix `AttributeError: 'ResourceVariable' object has no attribute '__pos__'` by returning `self.value` in `__pos__` in the TF backend.

You can fix the others by outright removing the `divmod`, rdivmod`, and `round` method. They just aren't very useful (hence why the array class of most backends don't implement them).

**Comment by Faisal-Alsrheed** [comment]
+ Removed __div__ and __rdiv__ methods.
These methods were related to Python 2 division operations and are not supported in Python 3.

**Comment by Faisal-Alsrheed** [comment]
Done @fchollet 
Thank you for your feedback.

**Pull request merged**